jupyter_server/terminal/handlers.py

```diff
@@ -27,13 +33,32 @@ def get(self, *args, **kwargs):
     def on_message(self, message):
         super(TermSocket, self).on_message(message)
         self._update_activity()
+        self._set_state_busy()
 
     def write_message(self, message, binary=False):
         super(TermSocket, self).write_message(message, binary=binary)
-        self._update_activity()
+        message_seg = json.loads(message)
+        if not self._first_stdout and message_seg[0] == 'stdout':
+            # Record the first output to identify the terminal return
+            # It works well for jupyterhub-singleuser and should also work for other debian-based mirrors
+            # fixme: May fail if terminal is not properly separated with ':' or change user after connect
+            #        (Any change to the user, hostname or environment may render it invalid)
+            self._first_stdout = message_seg[1].split(':')[0].lstrip()
```

Comment:
This meets my needs, but may not be a good general approach
Maybe we need to raise an issue with terminado?

Reply:
As I understand it, it is terminal dependent, terminado is just reading from the `pty`: https://github.com/jupyter/terminado/blob/ab4ced0b77c8f198cdc27fe1d52c523635083461/terminado/websocket.py#L77

Reply:
Yes, it may be a little difficult to implement a reliable solution based on the current architecture. But we can delve into how to determine that an terminal is running